### DISCUSSION TITLE
Conditional targets

### DISCUSSION THREAD

**Discussion by liutiming** [ropensci/targets#520] [Ideas]
Sometimes when I apply the same target workflow to another dataset, there will be some branches (e.g. `tar_map`) that return errors because the previous target returns nothing (e.g. because that step does some filtering). 

Is it therefore possible that we have targets that will run only if a certain conditions are met? Perhaps this can be an argument that takes in a function, and the function can be user-defined and contains `if` loops. If the function returns `FALSE`, then that target and all downstream targets will be removed from the workflow. 

Right now if I were to fix this issue on a per workflow basis, I will have to add that if loop and return a dummy dataframe for all downstream targets. A target-level solution seems to be more efficient, and can also avoid the more tricky cases when there is branching.

**Comment by wlandau** [comment]
This sort of flexibility is one of the primary goals of dynamic branching. You can create an upstream target to branch over in order to decide on the branches while the pipeline is running. This is the pattern I recommend in the majority of cases. There's `tar_cancel()` as an alternative, but that assumes you already have data on the branches you skip: https://github.com/ropensci/targets/discussions/519

**Reply by liutiming** [reply]
Thanks a lot!

**Comment by gorkang** [comment]
I am trying to understand how to create a conditional target with dynamic branching in the following situation.

We have regular files and sensitive files from an experimental protocol. The regular files are stored in 'data/' and the sensitive files in '.vault/data'. The end goal is to upload to Github the full data preparation and analysis R project, excluding the sensitive data (the '.vault/' folder won't be uploaded).

```
# Read and prepare regular files
tar_target(input_files, list.files(path = paste0("data/")), format = "file"),
tar_target(DF_raw, read_data(input_files)),

# Regular data
tar_target(df_REGULAR, prepare_REGULAR(DF_raw)),

# Sensitive data
  # For the sensitive data we use read_data inside with ".vault/data" files. 
  # We don't want any of the intermediate steps to end up in _targets.
  # The df_SENSITIVE is stripped of all sensitive data in prepare_SENSITIVE()
tar_target(df_SENSITIVE, prepare_SENSITIVE()), 

# DF ready for analysis
tar_target(DF_joined, create_joined(df_REGULAR, df_SENSITIVE)),

...
```

We want anyone to be able to run the project, so we need _targets to use the 'df_SENSITIVE' target output uploaded to the repo when the RAW sensitive data won't be available.

Our solution has been to include a conditional inside `prepare_SENSITIVE()`

```
prepare_SENSITIVE <- function() {

  if (length(list.files(".vault/data")) > 0) {
    # Process the RAW files
    df_SENSITIVE = ...
  } else if (file.exists(here::here("_targets/objects/df_SENSITIVE"))) {
    # No ".vault/data" available, just read df_SENSITIVE from _targets/objects
    df_SENSITIVE = read_rds(here::here("_targets/objects/df_SENSITIVE"))
  }
  return(df_SENSITIVE)

}
```

I am wondering if there is a cleaner way to solve this in the targets pipeline instead.

**Reply by wlandau** [reply]
So collaborators need to be able to run the full pipeline even when not all the upstream data is available to them? If you commit the data store (assuming that part is not as confidential as `.vault/data`) then you might be able to use #524, e.g. `tar_make(names = -any_of("df_SENSITIVE"), shortcut = TRUE)`.

But otherwise it does not seem feasible for others to run all parts of the pipeline  if they do not have access to all the required upstream data. Workarounds are possible though: for example, you could begin the pipeline from a less sensitive version of the data (maybe preprocessed in some way) and then share that as the full pipeline.
